src/v/cloud_storage/materialized_segments.cc

```diff
@@ -99,7 +103,7 @@ size_t materialized_segments::max_readers() const {
 
 size_t materialized_segments::max_segments() const {
     return static_cast<size_t>(_max_segments_per_shard().value_or(
-      _max_partitions_per_shard() * default_segment_factor));
+      _cache_size / _log_segment_size() * default_segment_factor));
```

Comment:
Is the goal here to generally reduce the amount of consumed memory? Or is it to enforce a specific bound on the number of files we keep around in our cache node-wide? If the latter, should we be adjusting by the number of cores?